File: FullStackApplications/LUPUS-IN-TABULA/frontend/domain/Lobby.js
import { Player } from "./Player";

export class Lobby {
  constructor({
    id,
    name,
    creator_id = null,
    status = "WAITING",
    min_players = 4,
    active_session_id = null,
    players = [],
    spectators = [],
  }) {
    this._id = id;
    this._name = name;
    this._creatorId = creator_id;
    this._status = status;
    this._minPlayers = min_players;
    this._activeSessionId = active_session_id;

    this._players = players.map((p) => (p instanceof Player ? p : new Player(p)));
    this._spectators = spectators.map((p) => (p instanceof Player ? p : new Player(p)));
  }

  id() {
    return this._id;
  }

  name() {
    return this._name;
  }

  creatorId() {
    return this._creatorId;
  }

  status() {
    return this._status;
  }

  minPlayers() {
    return this._minPlayers;
  }

  activeSessionId() {
    return this._activeSessionId;
  }

  players() {
    return [...this._players];
  }

  spectators() {
    return [...this._spectators];
  }

  hasPlayer(playerId) {
    return this._players.some((p) => p.id() === playerId);
  }

  hasSpectator(playerId) {
    return this._spectators.some((p) => p.id() === playerId);
  }

  isEmpty() {
    return this._players.length === 0 && this._spectators.length === 0;
  }

  withPlayer(player) {
    const p = player instanceof Player ? player : new Player(player);
    if (this.hasPlayer(p.id())) return this;
    return new Lobby({
      id: this._id,
      name: this._name,
      creator_id: this._creatorId,
      status: this._status,
      min_players: this._minPlayers,
      active_session_id: this._activeSessionId,
      players: [...this._players, p],
      spectators: this._spectators,
    });
  }

  withoutPlayer(playerId) {
    if (!this.hasPlayer(playerId)) return this;
    return new Lobby({
      id: this._id,
      name: this._name,
      creator_id: this._creatorId,
      status: this._status,
      min_players: this._minPlayers,
      active_session_id: this._activeSessionId,
      players: this._players.filter((p) => p.id() !== playerId),
      spectators: this._spectators,
    });
  }

  withSpectator(player) {
    const p = player instanceof Player ? player : new Player(player);
    if (this.hasSpectator(p.id())) return this;
    return new Lobby({
      id: this._id,
      name: this._name,
      creator_id: this._creatorId,
      status: this._status,
      min_players: this._minPlayers,
      active_session_id: this._activeSessionId,
      players: this._players,
      spectators: [...this._spectators, p],
    });
  }

  withoutSpectator(playerId) {
    if (!this.hasSpectator(playerId)) return this;
    return new Lobby({
      id: this._id,
      name: this._name,
      creator_id: this._creatorId,
      status: this._status,
      min_players: this._minPlayers,
      active_session_id: this._activeSessionId,
      players: this._players,
      spectators: this._spectators.filter((p) => p.id() !== playerId),
    });
  }
}

File: FullStackApplications/LUPUS-IN-TABULA/backend/app/domain/lobby.py
from app.domain.player import Player

class Lobby:
  def __init__(self, lobby_id: str, name: str, creator_id: str, min_players: int = 4):
    self._id = lobby_id
    self._name = name
    self._creator_id = creator_id
    self._status = "WAITING"
    self._min_players = min_players
    self._active_session_id = None
    self._players: list[Player] = []
    self._spectators: list[Player] = []

  def id(self) -> str:
    return self._id

  def name(self) -> str:
    return self._name

  def creator_id(self) -> str:
    return self._creator_id

  def status(self) -> str:
    return self._status

  def min_players(self) -> int:
    return self._min_players

  def active_session_id(self) -> str | None:
        return self._active_session_id

  def players(self) -> list[Player]:
    return list(self._players)

  def spectators(self) -> list[Player]:
    return list(self._spectators)

  def has_player(self, player_id: str) -> bool:
    return any(p.id() == player_id for p in self._players)

  def has_spectator(self, player_id: str) -> bool:
    return any(p.id() == player_id for p in self._spectators)

  def add_player(self, player: Player):
    if self.has_player(player.id()):
      return
    self._players.append(player)

  def add_spectator(self, player: Player):
    if self.has_spectator(player.id()):
      return
    self._spectators.append(player)

  def remove_member(self, player_id: str):
    self._players = [p for p in self._players if p.id() != player_id]
    self._spectators = [p for p in self._spectators if p.id() != player_id]

  def start(self):
    if self._status != "WAITING":
      raise ValueError("Lobby is not in WAITING")
    if len(self._players) < self._min_players:
      raise ValueError("Not enough players to start")
    self._status = "RUNNING"
    self._active_session_id = f"s-{self._id}"

  def end(self):
    if self._status != "RUNNING":
      raise ValueError("Lobby is not in RUNNING")
    self._status = "ENDED"

  def rematch(self):
    if self._status != "ENDED":
        raise ValueError("Lobby is not in ENDED")

    for s in self._spectators:
        if not self.has_player(s.id()):
            self._players.append(s)

        self._spectators = []
        self._status = "WAITING"
        self._active_session_id = None
